contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -20,54 +56,135 @@ mod test {
+        let store = MockStore::new();
+        let config = Config {
+            nexus: Addr::unchecked("nexus"),
+            router: Addr::unchecked("router"),
+        };
+        let contract = Contract { store, config };
+
+        let res = contract.route_messages(Addr::unchecked("unauthorized"), vec![]);
+
+        assert!(res.is_err_and(|err| matches!(err.current_context(), ContractError::Unauthorized)));
+    }
+
+    #[test]
+    fn route_to_nexus_with_no_msg() {
+        let store = MockStore::new();
+        let config = Config {
+            nexus: Addr::unchecked("nexus"),
+            router: Addr::unchecked("router"),
+        };
+        let contract = Contract { store, config };
+
+        let res = contract.route_messages(Addr::unchecked("router"), vec![]);
+
+        assert!(res.is_ok_and(|res| res.messages.is_empty()));
+    }
+
+    #[test]
+    fn route_to_nexus_with_msgs_that_have_not_been_routed() {
+        let mut store = MockStore::new();
+        let config = Config {
+            nexus: Addr::unchecked("nexus"),
+            router: Addr::unchecked("router"),
+        };
+        store
+            .expect_is_message_routed()
+            .times(2)
+            .returning(|_| Ok(false));
+        store
+            .expect_set_message_routed()
+            .times(2)
+            .returning(|_| Ok(()));
+        let contract = Contract { store, config };
 
-        assert!(res.is_ok_and(|res| res.messages.is_empty()))
+        let msgs = vec![
+            Message {
+                cc_id: CrossChainId {
+                    chain: "sourceChain".parse().unwrap(),
+                    id: "0x2fe4:0".parse().unwrap(),
+                },
+                source_address: "0xb860".parse().unwrap(),
+                destination_address: "0xD419".parse().unwrap(),
+                destination_chain: "destinationChain".parse().unwrap(),
+                payload_hash: decode(
+                    "bb9b5566c2f4876863333e481f4698350154259ffe6226e283b16ce18a64bcf1",
+                )
+                .unwrap()
+                .into(),
+            },
+            Message {
+                cc_id: CrossChainId {
+                    chain: "sourceChain".parse().unwrap(),
+                    id: "0x6b33:10".parse().unwrap(),
+                },
+                source_address: "0x0725".parse().unwrap(),
+                destination_address: "0x7FAD".parse().unwrap(),
+                destination_chain: "destinationChain".parse().unwrap(),
+                payload_hash: decode(
+                    "bb9b5566c2f4876863333e481f4698350154259ffe6226e283b16ce18a64bcf1",
+                )
+                .unwrap()
+                .into(),
+            },
+        ];
+        let res = contract.route_messages(Addr::unchecked("router"), msgs);
+
+        assert!(res.is_ok_and(|res| res.messages.len() == 2));
     }
 
     #[test]
-    fn test_route_to_nexus_with_msgs() {
+    fn route_to_nexus_with_msgs_that_have_been_routed() {
```

Comment:
might be good to test a mix of messages that have been routed and some that haven't